src/predictor/predict_fn.h

```diff
@@ -22,7 +22,7 @@ inline XGBOOST_DEVICE bst_node_t GetNextNode(const RegTree::Node &node, const bs
                  ? node.LeftChild()
                  : node.RightChild();
     } else {
-      return node.LeftChild() + !(fvalue < node.SplitCond());
+      return  fvalue < node.SplitCond() ? node.LeftChild() : node.RightChild();
```

Comment:
This change will lead to degraded performance and therefore cannot be accepted. Assuming that the right child node is +1 offset from the left child saves one function call and leads to up to 35% performance improvement (https://github.com/dmlc/xgboost/pull/6127).

In general, users are able to generate arbitrary JSON payload that will lead to undefined behavior when loaded by XGBoost. We cannot protect against all possible failure modes arising from this use case. Users should only load from JSON files that originate from the only trusted source: `Booster.save_model()`.

Reply:
Cool cool
I'll note that I don't think the 35% is over this version, but over an older version that is a lot different in terms of output assembly.
There's also an answered question by hcho3 in the pull request regarding the performance of exactly the change in this pull request.
Basically on any modern compiler and architecture the output assembly of this line should be a conditional move and not a branch jump.